### PR TITLE
Add general k support for topk sampling

### PR DESCRIPTION
**Pull request by smallv0221**
### PR types

New features
### PR changes

APIs 
### Description

Add general k support for topk sampling

### CODE REVIEW

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu [hunk 1]
```diff
@@ -553,7 +595,7 @@ void topK_sampling_kernel_kernelLauncher(void* workspace,
                                          DecodingSamplingArguments args,
                                          cudaStream_t stream) {
   std::minstd_rand engine;
-  int seed = std::random_device()();
+  int seed = 123;
```
Comment (ZeyuChen) [inline paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu]: Why 123 here？ Seems too tricky here.

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu [hunk 2]
```diff
@@ -65,6 +65,48 @@ __launch_bounds__(THREADBLOCK_SIZE) __global__
   }
 }
 
+template <typename T, int THREADBLOCK_SIZE>
+__global__ void beam_topK_kernel_general(const T* log_probs,
```
Comment (ZeyuChen) [inline paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu:69]: This function name is not easy to understand. Adjective word at last is not a good idea.

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu (hunk 1 above)
Comment (smallv0221) [reply paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu]: This pr is still working in progress. The seed is for testing.

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu (hunk 2 above)
Comment (smallv0221) [reply paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu:69]: 我是跟FT的命名规范学的，他本身就有`topk_stage_1_opt3`和`topk_stage_1_opt2_general`这种命名方式……就跟它保持一致吧

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu (hunk 1 above)
Comment (smallv0221) [reply paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu]: Done

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu [hunk 3]
```diff
@@ -563,25 +605,40 @@ void topK_sampling_kernel_kernelLauncher(void* workspace,
...
-    int* topk_tmp_id_buf = (int*)workspace;
+    T* temp_log_probs = (T*)workspace;
+    int* topk_tmp_id_buf = (int*)(temp_log_probs + temp_log_probs_buf_size);
     T* topk_tmp_val_buf = (T*)(topk_tmp_id_buf + topk_tmp_ids_buf_size);
 
     switch (candidate_num) {
       CASE_K(1);
       CASE_K(2);
-      CASE_K(4);
+      CASE_K(4)
```
Comment (FrostML) [inline paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu]: missing `;` ?

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu (hunk 2 above)
Comment (ZeyuChen) [reply paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu:69]: ok

**Review by ZeyuChen — COMMENTED** [review]
LGTM

File: paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu [hunk 4]
```diff
@@ -563,25 +605,40 @@ void topK_sampling_kernel_kernelLauncher(void* workspace,
...
   topk_tmp_ids_buf_size = (int)(ceil(topk_tmp_ids_buf_size / 4.)) * 4;
+
   topk_tmp_val_buf_size = (int)(ceil(topk_tmp_val_buf_size / 4.)) * 4;
 
   if (workspace == nullptr) {
-    workspace_size = sizeof(int) * topk_tmp_ids_buf_size +
-                     sizeof(int) * topk_tmp_val_buf_size;
+    workspace_size = sizeof(float) * temp_log_probs_buf_size +
+                     sizeof(int) * topk_tmp_ids_buf_size +
+                     sizeof(float) * topk_tmp_val_buf_size;
```
Comment (FrostML) [inline paddlenlp/ops/patches/FasterTransformer/cuda/topk_kernels.cu]: `float` or type `T`?

**Comment by FrostML** [comment]
麻烦再在 `float16` 下验证一下吧，实际可能会使用 `float16`。执行时加上 `--use_fp16_decoding`。

**Review by FrostML — DISMISSED** [review]
LGTM. @guoshengCS 看看还有没有什么意见

**Review by guoshengCS — APPROVED** [review]
LGTM

**Pull request merged**